### PR TITLE
Rework counter logic test for rule reloading

### LINKED ISSUE #67: Rework the counter logic test
In a [comment to a pull request](https://github.com/juspay/fencer/pull/51/files#r337047763), @neongreen suggested several changes to the test that checks the counter logic. Implement what has been suggested there.

### PR DESCRIPTION
**Pull request by mdimjasevic**
This patch reorganizes the code to more closely reflect what it is about (this involves renaming and moving functions around). Furthermore, it updates a rule reloading test that checks if limits remaining are updated as expected.

Closes #67.

### CODE REVIEW

**Comment by mdimjasevic** [comment]
@neongreen , just a gentle reminder that I could use your review for this PR.

File: lib/Fencer/Logic.hs [hunk 1]
```diff
@@ -2,20 +2,19 @@
 {-# LANGUAGE OverloadedLabels #-}
 {-# LANGUAGE RecordWildCards #-}
 
--- | In-memory state of Fencer.
-module Fencer.AppState
+-- | In-memory state of Fencer and the logic for managing it.
+module Fencer.Logic
     ( AppState
-    , appStateCounters
     , appStateRules
```
Comment (neongreen) [inline lib/Fencer/Logic.hs]: Can you also remove this export? It's not used anywhere else in the code and it allows breaking internal invariants.

File: test/Fencer/Logic/Test.hs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      st1 <- getRemainingLimit <$> makeAHit state
+      assertBool
+        "The remaining rate limit was not updated!"
+        (st1 == limit - hits)
+
+      -- Set the new rules and the rules reloaded flag
+      atomically $ setRules state (mapRuleDefs definitions2)
+      -- Record a hit and get a remaining limit
+      st2 <- getRemainingLimit <$> makeAHit state
+      assertBool
```
Comment (neongreen) [inline test/Fencer/Logic/Test.hs]: You should use `assertEqual` to get better errors from tasty when the test fails.

I pushed a commit that adds an HLint rule to this effect, you can do `hlint --git` to see it in action.

File: test/Fencer/Logic/Test.hs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+-- | Test that a rule limit unit change adds a new counter and leaves
+-- the old one intact.
+test_logicLimitUnitChange :: TestTree
+test_logicLimitUnitChange =
+  withServer $ \serverIO ->
+    testCase "A rule limit unit change on rule reloading" $ do
+      state <- serverAppState <$> serverIO
+      void $ atomically $ setRules state (mapRuleDefs definitions1)
+
+      -- Record a hit and get a remaining limit
```
Comment (neongreen) [inline test/Fencer/Logic/Test.hs]: Not 100% sure, but:

```suggestion
      -- Record a hit and get the remaining limit
```

File: test/Fencer/Logic/Test.hs (hunk 2 above)
Comment (mdimjasevic) [reply test/Fencer/Logic/Test.hs]: A good point! Somehow this slipped through. Took care of it in an upcoming commit.

Haven't checked, but it'd be great to have `hlint` run as a Git hook or something similar.

File: lib/Fencer/Logic.hs (hunk 1 above)
Comment (mdimjasevic) [reply lib/Fencer/Logic.hs]: I guess I introduced it at some point and forgot to remove it. Thank you for spotting it!

This is taken care of in an upcoming commit.

File: test/Fencer/Logic/Test.hs (hunk 2 above)
Comment (neongreen) [reply test/Fencer/Logic/Test.hs]: You can add a hook locally + add instructions for how to do so to CONTRIBUTING.md. 

HLint also runs on Code Climate. It used to ignore `tests/` but starting from the next commit it should work on `tests/` too.

Comment (neongreen) [reply test/Fencer/Logic/Test.hs]: https://codeclimate.com/github/juspay/fencer/pull/69

**Comment by neongreen** [comment]
> Commits:
> ...
> Issue #32: create contribution guidelines (#62)
> Issue #68: fix the branch name template rendering (#72)

Something got wrong with Git. You might want to rebase on top of latest master.

**Comment by mdimjasevic** [comment]
> > Commits:
> > ...
> > Issue #32: create contribution guidelines (#62)
> > Issue #68: fix the branch name template rendering (#72)
> 
> Something got wrong with Git. You might want to rebase on top of latest master.

I might have marked incorrect issues in my commit messages. This shouldn't matter as the commits will all be squashed upon merging.

File: test/Fencer/Logic/Test.hs (hunk 2 above)
Comment (mdimjasevic) [reply test/Fencer/Logic/Test.hs]: > https://codeclimate.com/github/juspay/fencer/pull/69

Fixed.

**Review by neongreen — APPROVED** [review]

**Pull request merged**